src/pudl/etl/epacems_assets.py

```diff
@@ -148,8 +149,438 @@ def hourly_emissions_epacems(
+        is the case when there are units that report to CEMS but which do not exist in
+        the EIA data, we assign these units to a single subplant.
+
+    Args:
+        subplant_crosswalk: a dataframe containing the output of
+            :func:`pudl.analysis.epacamd_eia.make_subplant_ids`
+    """
+    # Step 1: Create corrected versions of subplant_id and unit_id_pudl
+    # if multiple unit_id_pudl are connected by a single subplant_id, unit_id_pudl_connected groups these unit_id_pudl together
+    subplant_crosswalk = connect_ids(
+        subplant_crosswalk, id_to_update="unit_id_pudl", connecting_id="subplant_id"
+    )
+    # if multiple subplant_id are connected by a single unit_id_pudl, group these subplant_id together
+    subplant_crosswalk = connect_ids(
+        subplant_crosswalk, id_to_update="subplant_id", connecting_id="unit_id_pudl"
+    )
+
+    # Step 2: Fill missing subplant_id
+    # We will use unit_id_pudl to fill missing subplant ids, so first we need to fill any missing unit_id_pudl
+    # We do this by assigning a new unit_id_pudl to each generator that isn't already grouped into a unit
+
+    # create a numeric version of each generator_id
+    # ngroup() creates a unique number for each element in the group
+    subplant_crosswalk["numeric_generator_id"] = subplant_crosswalk.groupby(
+        ["plant_id_eia", "generator_id"], dropna=False
+    ).ngroup()
+    # when filling in missing unit_id_pudl, we don't want these numeric_generator_id to overlap existing unit_id
+    # to ensure this, we will add 1000 to each of these numeric generator ids to ensure they are unique
+    # 1000 was chosen as an arbitrarily high number, since the largest unit_id_pudl is ~ 10.
+    subplant_crosswalk["numeric_generator_id"] = (
+        subplant_crosswalk["numeric_generator_id"] + 1000
+    )
+    # fill any missing unit_id_pudl with a number for each unique generator
+    subplant_crosswalk["unit_id_pudl_filled"] = (
+        subplant_crosswalk["unit_id_pudl_connected"]
+        .fillna(subplant_crosswalk["subplant_id_connected"] + 100)
+        .fillna(subplant_crosswalk["numeric_generator_id"])
+    )
+    # create a new unique subplant_id based on the connected subplant ids and the filled unit_id
+    subplant_crosswalk["new_subplant"] = subplant_crosswalk.groupby(
+        ["plant_id_eia", "subplant_id_connected", "unit_id_pudl_filled"],
+        dropna=False,
+    ).ngroup()
+
+    return subplant_crosswalk
+
+
+def connect_ids(df, id_to_update, connecting_id):
+    """Corrects an id value if it is connected by an id value in another column.
+
+    If multiple subplant_id are connected by a single unit_id_pudl, this groups these
+    subplant_id together. If multiple unit_id_pudl are connected by a single subplant_id,
+    this groups these unit_id_pudl together.
+
+    Args:
+        df: dataframe containing columns with id_to_update and connecting_id columns
```

Comment:
Might make the function purpose clearer if we rename `df` to something less generic.